### DISCUSSION TITLE
Everybody is talking about replacing the Mainboard, but what about just adding an RPI3 or 4?

### DISCUSSION THREAD

**Discussion by chrisz99** [xblax/flashforge_ad5m_klipper_mod#254] [General]
I'm thinking about adding an RPI 3 or 4 to the back of the printer, connecting the TFT cable from the existing LCD to the PI and running Fluid off of the PI in an attempt to mitigate ram usage off the mainboard. At least I'm assuming that's how other printers accomplish this type of setup anyways, and I'm assuming this should work.

**Comment by jeffalanmcclain** [comment]
Just FYI, I'm running a remote RPi5 and have disabled the Fluidd and Mainsail services on the printer and am running those services on the remote RPi5.  It was a PitA to figure out how to get everything configured, because Kiauh does NOT like to install just fluidd, octoprint, mainsail without a full Klipper/moonraker install.  So I did that, then manually reconfigured them to point at the moonraker running on the FFA5M and that has been great.  It has NOT, however, fixed the large number of TTC fails I tend to get sometimes, and I'm still looking for ways to migrate even more off the FFA5M controller.  TWO things would be to physically mount the RPi5 (like you mention) and run the Klipperscreen function from it as well as the USB camera feed.  I'm still not sure how that will function, given there needs to be some level of Klipper (on printer) control on that machine for camera capture/timelapse...still thinking this through...

I'm still hoping/wishing for an update on the KlipperMOD files to incorporate some of the newer releases of moonraker/klipper (stuff like KAMP isn't supported on the KlipperMOD currently here)...  I also can't have moonraker software updates running or it will try to update klipperMOD files.

P.S. this might just be a short term, as the MCU timer too close failures have just about got me loosing my mind and I'm seriously considering building a 350mm Voron Trident and just abandoning this 5M.  It has been GREAT for learning and getting into 3D Printing and actually prints amazingly, but since I've been running KlipperMOD, it just limits too many things I can do on Klipper threads.  I'm running a custom Knomi2 I installed in the print head, spoolman all just fine as well as second bed/chamber air curtain fan on right side successfully, however, I want to do an ERCF filament switcher and KAMP and stuff just keeps seeming limited...

**Reply by chrisz99** [reply]
I have a similar setup to you, except im running a RPI 3B , and I had no issues using KIUAH installer to just install the fluid interface. I disabled the mainsail, and fluid interfaces like you have with the help of your guide I believe. I then installed a custom cloudflared service to forward my local ips for the camera server and the Moonraker server so I can connect to it off of my wifi. 

Like you I still have TTC errors, however they are very very infrequent. I just had one this morning after boot, I think I just uploaded a print file and it crashed.

I explored running klipper screen and the camera off the rpi directly with the existing hardware and determined that it might be impossible. RPI uses 0.5mm pitch FPC cables while, the 5m uses 1mm pitch FPC cabling. Also the MCU on the display runs the Moonraker instance and communicates with the main board. If you can't get the RPI to communicate with the main board, you can't remove the stock display.

With that being said, I looked into replacing the main board. That's difficult because of all the proprietary hardware, for example the bed connections are a ribbon cable which is atypical. As well as the toolhead main board connection to the board ( 6 pin connection ), although there are more workarounds with that one.

Idk why flashforge decided to go this route with all the proprietary hardware. I love the 5m pro I have, I just wish I could change out the main board, and whatever else I want too. 

As far as Kamp goes, Kamp works with this klipper mod, just follow installation as normal. I haven't gotten adaptive purging to work though, so be careful with adaptive meshing and the default line purging ( nozzle scraping bed ). The bed meshing python file is located in klippy/extras, I'm theorizing you can just replace the python file with the updated bed mesh with adaptive bed meshing from klipper 0.12, and get those features. But I have not tried that yet.

**Reply by jeffalanmcclain** [reply]
Very cool.  I may need to pick your brain more on trying to get remote moonraker working (then I might have some luck with moving at least the camera over).  Sounds like we are in similar camps.  I love how my 5M prints too, however I'm just fed up with the TTC fails.  Done.  That and to a lesser extent, I would like to do more on a filament changer and it just feels like I'm pushing the existing MCU beyond it's limits.  I've had 5 TTC for ONE good print today.  I guess I should go look at Kamp install more.  I hesitated to do it, just because I don't want to even touch the KlipperMOD install on the printer incase something is incompatible.  I guess I meant the localized adaptive bed mesh you can run with the latest Klipper (like KAMP brings) doesn't work on our current 0.5 KlipperMOD version.  I hadn't wanted to try replacing individual python files with that updated adaptive from 0.12, because it makes me feel like it could be razors edge on causing more problems...

I agree with you on the futility of mainboard replacement.  I went down that path investigating too.  It honestly just convinced me more that I need to just build a Voron Trident.  My biggest fear with that is that I invest $1500 and a LOT of time building a nice new printer that might STILL have TTC errors (I've heard CB2 still have some, so would probably do CM4) and/or might not print as nice as I have the 5M Pro tuned up...

I put in a Knomi v2 in a new print cover I designed and it works awesome on this.  I should probably upload the STL files for it on printables... *edit* here it is: https://www.printables.com/model/1046164-knomi-fan-shroud-for-flashforge-adventurer-5m

**Reply by chrisz99** [reply]
So I've confirmed that replacing the bedmesh.py file with Klipper 0.12 bedmesh.py file does indeed work, and does add native Adaptive Bed Meshing to the klipper firmware. I find that it works slightly better than KAMP, in the way that it doesn't scrape / has better calibration line at areas of the bed that are not probed.

Honestly man, it seems like it would be more worthwhile for us to build a custom printer like a Voron as you mentioned. Or switch to something more open source and less proprietary such as the Creality K1C.

 I own an Ender 5 plus that's modded to high water, and one thing about creality products is that there hardware is the same / similiar to open source hardware. I can throw / slap any part on my Ender 5 plus, and make it work. The mainboard is an open book, and is also easily replaceable. I modded my E5 plus to Core XY and linear rails , mainboard swap, extruder swap, the whole kit and kaboodle. The K1C seems to be the same way, the mainboard is similiar to the Ender 5 / Ender 3, and silent boards from them. I reckon you can swap that board with anything you desire too, aswell as changing the host mcu.

**Reply by rEkCaRc1337** [reply]
> Very cool. I may need to pick your brain more on trying to get remote moonraker working (then I might have some luck with moving at least the camera over). Sounds like we are in similar camps. I love how my 5M prints too, however I'm just fed up with the TTC fails. Done. That and to a lesser extent, I would like to do more on a filament changer and it just feels like I'm pushing the existing MCU beyond it's limits. I've had 5 TTC for ONE good print today. I guess I should go look at Kamp install more. I hesitated to do it, just because I don't want to even touch the KlipperMOD install on the printer incase something is incompatible. I guess I meant the localized adaptive bed mesh you can run with the latest Klipper (like KAMP brings) doesn't work on our current 0.5 KlipperMOD version. I hadn't wanted to try replacing individual python files with that updated adaptive from 0.12, because it makes me feel like it could be razors edge on causing more problems...
> 
> I agree with you on the futility of mainboard replacement. I went down that path investigating too. It honestly just convinced me more that I need to just build a Voron Trident. My biggest fear with that is that I invest $1500 and a LOT of time building a nice new printer that might STILL have TTC errors (I've heard CB2 still have some, so would probably do CM4) and/or might not print as nice as I have the 5M Pro tuned up...
> 
> I put in a Knomi v2 in a new print cover I designed and it works awesome on this. I should probably upload the STL files for it on printables... _edit_ here it is: https://www.printables.com/model/1046164-knomi-fan-shroud-for-flashforge-adventurer-5m

I'm using Prusa Slicer and is kinda doing adaptive purging, just via gcode and the slicer instead of the printer itself:
```
START_PRINT BED_TEMP={first_layer_bed_temperature[0]} EXTRUDER_TEMP={first_layer_temperature[0]} DISABLE_PRIMING=true
G1 Z5 F5000 ; lift nozzle
G92 E0.0 ; Reset Extruder
G1 X{first_layer_print_min[0]-5} Y{first_layer_print_min[1]} Z0.8 F6000.0
G1 X{first_layer_print_min[0]-5} Y{first_layer_print_min[1]+30} E30 F240.0
G92 E0
G1 E-0.5 F2100
G1 Y{first_layer_print_min[1]+40} F6000.0
G92 E0 ; Reset Extruder
```

Was using this before on my prior printers too, should be possible to do this in orca too if you're using that, i think the only slicer which might not support this parametric location for purge line is Cura.

![image](https://github.com/user-attachments/assets/ba862ee2-f0de-4fab-afb5-12e259ca1bdf)
